feat: 🎸 Add SQFormMultiValue component

20BBrown14 opened the pull request:
Add new component, SQFormMultiValue

Loom: https://www.loom.com/share/e2405d5fbe1a42b7a7d0b457132c5593

Sorry for the loom length :/ the last half is just code, though, if you care about that.

✅ Closes: #303

**When this is approved I'll add an issue to add tests for this component**

![Screen Shot 2021-06-17 at 8 57 34 AM](https://user-images.githubusercontent.com/29528409/122411112-1c0bd880-cf4a-11eb-8169-21e08dd54796.png)

hvilander commented:
Listening to your loom, what happens if I type in a string that is equal to the `value` of an option.

Example: options = { red: 1, blue: 2} and I type in 2.

20BBrown14 commented:
> Listening to your loom, what happens if I type in a string that is equal to the `value` of an option.
> 
> Example: options = { red: 1, blue: 2} and I type in 2.

Good question. So for your example, nothing. BEcause 2 and "2" are different. But if you had { red: 'color-red', blue: 'color-blue' } and you typed in "color-blue" you'll end up with just "Blue" in the options

seen here: https://www.loom.com/share/8f331a81f57c4bf5bb533cbc0c03d126

hvilander commented on src/components/SQForm/SQFormMultiValue.js:
```diff
@@ -0,0 +1,271 @@
+    const LIST_OVERSCAN_COUNT = 5;
+    const items = React.Children.toArray(children);
+    const ITEM_COUNT = items.length;
+    const ITEM_SIZE = 36;
```
Should these constants live with the one on line 10?

SeanGroff replied:
These constants are only used in this closure so they were scoped to this closure

Chris-Boyle commented on stories/SQForm.stories.js:
```diff
@@ -197,6 +211,13 @@ export const BasicForm = () => {
+        <SQFormMultiValue
+          name="favoriteColors"
+          label="Your Favorite Colors"
+          size={12}
```
Size 12 is a little extreme.  I don't think we will be doing that large of a field in any of our apps so we probably shouldn't do that in our Storybook as well.  We want to keep the stories as close to real world examples as possible.

SeanGroff replied:
+1
I've never realized how wide that form gets if the right side bar in Storybook is hidden.
Branden all up in here with the ultra wide lol.

Can we make this a 4 or something? To try and get a realistic view of how it'll behave in our apps.
Or maybe we add a max-width to the form wrapper so it doesn't grow so wide.

20BBrown14 replied:
![Screen Shot 2021-06-17 at 8 57 34 AM](https://user-images.githubusercontent.com/29528409/122411065-13b39d80-cf4a-11eb-86f2-7aa5b69d911d.png)

Chris-Boyle commented on stories/SQForm.stories.js:
```diff
@@ -29,7 +29,8 @@ import {
   SQFormRadioButtonGroup,
   SQFormCheckboxGroup,
   SQFormMaskedTextField,
```
[nit] move `SQFormMultiValue` above the MASKS to make the imports look more alphabetical. (I know they all aren't perfect)

Chris-Boyle commented:
What happens if your selected options take up more width than the field?

SeanGroff commented on package.json:
```diff
@@ -134,7 +134,7 @@
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
-    "react-scripts": "^4.0.0",
+    "react-scripts": "^3.4.3",
```
Blocker - we lose things like React Fast Refresh with this downgrade.
We have full control of our linter. Fix the specific issue, preferably in another PR.
No need for this to get in the way of the SQFormMultiValue PR

SeanGroff commented on src/components/SQForm/SQFormMultiValue.js:
```diff
@@ -0,0 +1,271 @@
+      );
+    }, [ITEM_COUNT, items]);
+
+    const getItemSize = React.useCallback(() => ITEM_SIZE, []);
```
I know this might be copy/pasta from the MUI docs example, but if this only returns `ITEM_SIZE` does this need to be a function?

vs passing `ITEM_SIZE` to the `itemSize` prop on line 59?

20BBrown14 replied:
tbh idk, i copied this from the autocomplete component. i can try to clean it up though.

20BBrown14 replied:
Turns out this does need to be a function, actually. `getItemSize` is being passed to `VariableSizeList` from react-window for the `itemSize` prop which requires a function.

SeanGroff commented on stories/SQForm.stories.js:
```diff
@@ -299,6 +328,14 @@ export const FormWithValidation = () => {
+        <SQFormMultiValue
+          name="favoriteColors"
+          label="Your Favorite Colors"
+          size={12}
```
Storybook is great for what I call UI chaos testing.
I question what happens if we change this to size `6` then select all of the options. What happens?

20BBrown14 replied:
Good question And so is

> What happens if your selected options take up more width than the field?

It wraps the options and grows vertically when the options are more than allowed in the width. if a single option is longer than the width of the field it truncates the option label.

![Screen Shot 2021-06-17 at 8 51 02 AM](https://user-images.githubusercontent.com/29528409/122410069-3c876300-cf49-11eb-9c05-0187f8bf9a1a.png)

Chris-Boyle replied:
@20BBrown14  what is that little x all by itself on the right there?  I didn't see that in the Loom.  Does that clear all the fields?  
![image](https://user-images.githubusercontent.com/20329437/122410520-9ab44600-cf49-11eb-9b03-14abe2c53504.png)

20BBrown14 replied:
I pointed that out in the loom as a way to clear the field at about 43 seconds

SeanGroff replied:
> I pointed that out in the loom as a way to clear the field at about 43 seconds
> 
> edit: and is also how SQFormAutocomplete works so our users and devs should be familiar.

Yep. It's a great feature to have! I think you'll want to run the clear button by Nemeth. My hunch is he's going to want it to always stick to the bottom regardless of the number of rows so it's always in the same spot near the inputs bottom border. The default from MUI sort of looks likes its floating in no mans land.

20BBrown14 replied:
Nemeth wanted it on the bottom, this is what i was able to get 
![Screen Shot 2021-06-17 at 11 07 36 AM](https://user-images.githubusercontent.com/29528409/122434002-55e5da80-cf5c-11eb-9b29-1e83131790f8.png)
![Screen Shot 2021-06-17 at 11 07 22 AM](https://user-images.githubusercontent.com/29528409/122434004-567e7100-cf5c-11eb-826e-954651f1f34e.png)

Chris-Boyle commented:
We are missing component documentation for this component.  Adding the validation example and reasoning would be good in the documentation.

20BBrown14 commented:
> We are missing component documentation for this component. Adding the validation example and reasoning would be good in the documentation.

I'm not sure what you're referring. Is there some documentation for the components I wish i knew about several months ago?

Chris-Boyle commented:
We do notes on some components.  We used to do it a lot more.  I am not sure where the stance is but it is nice to have for things that need a little more explanation.
https://github.com/SelectQuoteLabs/SQForm/tree/master/notes
Here is an example
https://5f4431386ea00a00220d495c-qfuecrbuoq.chromatic.com/?path=/docs/components-sqformdatepicker--date-picker-before-2024

SeanGroff reviewed (approved):
Fantastic work @20BBrown14!

SeanGroff commented:
Great job building this component @20BBrown14 , accepting PR feedback, and turning the feedback into a polished solution 💯  

I stamped my approval, pending the removal of the package-lock.json that's still there.

20BBrown14 commented:
> We do notes on some components. We used to do it a lot more. I am not sure where the stance is but it is nice to have for things that need a little more explanation.
> https://github.com/SelectQuoteLabs/SQForm/tree/master/notes
> Here is an example
> https://5f4431386ea00a00220d495c-qfuecrbuoq.chromatic.com/?path=/docs/components-sqformdatepicker--date-picker-before-2024

Ah, that's fair. I added a notes page that explains some stuff. Validation would probably be the trickiest thing. I also added component stories, which I left out originally.

20BBrown14 commented:
Thanks for all the feedback and approvals. Merging since everything seems to be addressed 🎉 🌮
